Classes/Emogrifier.php

```diff
@@ -1289,41 +1292,71 @@ function (array $matches) {
         $trimmedLowercaseSelector = trim($lowercasePaddedSelector);
         $xPathKey = md5($trimmedLowercaseSelector);
         if (!isset($this->caches[self::CACHE_KEY_XPATH][$xPathKey])) {
-            $roughXpath = '//' . preg_replace(
-                array_keys($this->xPathRules),
-                $this->xPathRules,
-                $trimmedLowercaseSelector
-            );
-            $xPathWithIdAttributeMatchers = preg_replace_callback(
-                self::ID_ATTRIBUTE_MATCHER,
-                [$this, 'matchIdAttributes'],
-                $roughXpath
-            );
-            $xPathWithIdAttributeAndClassMatchers = preg_replace_callback(
-                self::CLASS_ATTRIBUTE_MATCHER,
-                [$this, 'matchClassAttributes'],
-                $xPathWithIdAttributeMatchers
-            );
+            $notSelector = null;
+            if (strstr($trimmedLowercaseSelector, ':not')) {
+                list($trimmedLowercaseSelector, $notSelector) = explode(':not', $trimmedLowercaseSelector);
+            }
 
-            // Advanced selectors are going to require a bit more advanced emogrification.
-            // When we required PHP 5.3, we could do this with closures.
-            $xPathWithIdAttributeAndClassMatchers = preg_replace_callback(
-                '/([^\\/]+):nth-child\\(\\s*(odd|even|[+\\-]?\\d|[+\\-]?\\d?n(\\s*[+\\-]\\s*\\d)?)\\s*\\)/i',
-                [$this, 'translateNthChild'],
-                $xPathWithIdAttributeAndClassMatchers
-            );
-            $finalXpath = preg_replace_callback(
-                '/([^\\/]+):nth-of-type\\(\s*(odd|even|[+\\-]?\\d|[+\\-]?\\d?n(\\s*[+\\-]\\s*\\d)?)\\s*\\)/i',
-                [$this, 'translateNthOfType'],
-                $xPathWithIdAttributeAndClassMatchers
-            );
+            $finalXpath = '//' . $this->translateCssToXpathPass($trimmedLowercaseSelector);
+
+            if ($notSelector) {
```

Comment:
$notSelector currently is either a string or null, not a boolean. I propose to always make it a string and then check for !== '' here.